File: src/gateway/userAgent.js
// userAgent — turn a User-Agent string into { device, os, browser }.
//
// The browser sends this header on every request for free (the beacon's
// fetch included). It's a messy, lie-prone string, but enough to label a
// client "macOS · Chrome" or "iOS · Safari" for the users table. Zero deps:
// just ordered regex, most-specific first (order matters — iPad claims to be
// Macintosh, Edge claims to be Chrome, so the narrow cases win).

export function parseUA(ua = "") {
  const s = String(ua);

  // ---- OS ----
  let os = "Unknown";
  if (/iPhone/i.test(s)) os = "iOS";
  else if (/iPad/i.test(s)) os = "iPadOS";
  else if (/Android/i.test(s)) os = "Android";
  else if (/Mac OS X|Macintosh/i.test(s)) os = "macOS";
  else if (/Windows NT/i.test(s)) os = "Windows";
  else if (/CrOS/i.test(s)) os = "ChromeOS";
  else if (/Linux/i.test(s)) os = "Linux";

  // ---- browser (order: narrow impostors before the thing they impersonate) ----
  let browser = "Unknown";
  if (/Edg\//i.test(s)) browser = "Edge";
  else if (/OPR\/|Opera/i.test(s)) browser = "Opera";
  else if (/Firefox\//i.test(s)) browser = "Firefox";
  else if (/Chrome\//i.test(s)) browser = "Chrome";
  else if (/Safari\//i.test(s)) browser = "Safari";
  else if (/curl\//i.test(s)) browser = "curl";
  else if (/node|axios|python-requests|Go-http|Java\//i.test(s)) browser = "script";
  else if (/bot|crawler|spider|slurp/i.test(s)) browser = "bot";

  // ---- device class ----
  let device = "Desktop";
  if (/Mobile|iPhone|Android.*Mobile/i.test(s)) device = "Mobile";
  else if (/iPad|Tablet/i.test(s)) device = "Tablet";
  else if (browser === "curl" || browser === "script" || browser === "bot") device = "Machine";

  return { device, os, browser, raw: s.slice(0, 160) };
}

File: internal/useragent/useragent.go
// Package useragent turns a User-Agent string into {device, os, browser}.
//
// The browser sends this header on every request for free (the beacon's fetch
// included). It is a messy, lie-prone string, but enough to label a client
// "macOS · Chrome" or "iOS · Safari" in the users table.
//
// Ordered matching, most specific first — the order is the whole trick, because
// an iPad claims to be a Macintosh and Edge claims to be Chrome, so the narrow
// impostor must be tested before the thing it impersonates.
package useragent

import "strings"

type Parsed struct {
	Device  string `json:"device"`
	OS      string `json:"os"`
	Browser string `json:"browser"`
	Raw     string `json:"raw"`
}

func contains(haystack string, needles ...string) bool {
	for _, n := range needles {
		if strings.Contains(haystack, n) {
			return true
		}
	}
	return false
}

func Parse(ua string) Parsed {
	s := strings.ToLower(ua)

	// ---- OS ----
	os := "Unknown"
	switch {
	case contains(s, "iphone"):
		os = "iOS"
	case contains(s, "ipad"):
		os = "iPadOS"
	case contains(s, "android"):
		os = "Android"
	case contains(s, "mac os x", "macintosh"):
		os = "macOS"
	case contains(s, "windows nt"):
		os = "Windows"
	// "cros" as a bare substring also matches "microsoft" — real UA strings such
	// as Microsoft-WebDAV-MiniRedir were being reported as ChromeOS. Chrome OS
	// always spells it "CrOS " followed by the architecture, so require the
	// token boundary.
	case contains(s, "cros "), strings.HasSuffix(s, "cros"), contains(s, "cros;"), contains(s, "cros)"):
		os = "ChromeOS"
	case contains(s, "linux"):
		os = "Linux"
	}

	// ---- browser (narrow impostors before the thing they impersonate) ----
	browser := "Unknown"
	switch {
	case contains(s, "edg/"):
		browser = "Edge"
	case contains(s, "opr/", "opera"):
		browser = "Opera"
	case contains(s, "firefox/"):
		browser = "Firefox"
	case contains(s, "chrome/"):
		browser = "Chrome"
	case contains(s, "safari/"):
		browser = "Safari"
	case contains(s, "curl/"):
		browser = "curl"
	case contains(s, "node", "axios", "python-requests", "go-http", "java/"):
		browser = "script"
	case contains(s, "bot", "crawler", "spider", "slurp"):
		browser = "bot"
	}

	// ---- device class ----
	device := "Desktop"
	switch {
	case contains(s, "mobile", "iphone"):
		device = "Mobile"
	case contains(s, "ipad", "tablet"):
		device = "Tablet"
	case browser == "curl" || browser == "script" || browser == "bot":
		device = "Machine"
	}

	raw := ua
	if len(raw) > 160 {
		raw = raw[:160]
	}
	return Parsed{Device: device, OS: os, Browser: browser, Raw: raw}
}
